### PR TITLE
OpenAPI: added host override and detection for potential loops

### PR DESCRIPTION
**Pull request by psiinon**
The change to only support 2.6.0 is deliberate - more than 1/2 of our users have upgraded now so I dont think we need to keep supporting old versions indefinitely. Esp for relatively minor enhancements.

### CODE REVIEW

File: src/org/zaproxy/zap/extension/openapi/resources/help/contents/openapi.html [hunk 1]
```diff
@@ -18,12 +18,15 @@
 <li>Import an Open API definition from the local file system</li>
 <li>Import an Open API definition from a URL</li>
 </ul>
+The URL Import dialog supports an optional additional field that allows you to override the Host (and port) in the
+definition with an alternative one that you supply.<br>
+This option is not provide for the file import as its easy to change a local file to use a different host.
```
Comment (kingthorin) [inline src/org/zaproxy/zap/extension/openapi/resources/help/contents/openapi.html]: `provided`

File: src/org/zaproxy/zap/extension/openapi/ImportFromUrlDialog.java [hunk 2]
```diff
@@ -111,7 +121,7 @@ public void actionPerformed(ActionEvent e) {
             String url = fieldURL.getText();
             /* Calls a parsing task in a new thread. */
             try {
-                caller.importOpenApiDefinition(new URI(url, false), true);
+                caller.importOpenApiDefinition(new URI(url, false), siteOverride.getText(), true);
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/openapi/ImportFromUrlDialog.java]: Worth validating that the provided host override is syntactically valid?

File: src/org/zaproxy/zap/extension/openapi/OpenApiAPI.java [hunk 3]
```diff
@@ -67,7 +68,8 @@ public ApiResponse handleApiAction(String name, JSONObject params) throws ApiExc
         } else if (ACTION_IMPORT_URL.equals(name)) {
 
             try {
-                List<String> errors = extension.importOpenApiDefinition(new URI(params.getString(PARAM_URL), false), false);
+                List<String> errors = extension.importOpenApiDefinition(new URI(params.getString(PARAM_URL), false),
+                        params.optString(PARAM_HOST_OVERRIDE, null), false);
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/openapi/OpenApiAPI.java]: Worth validating that the provided host override is syntactically valid?

File: src/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java [hunk 4]
```diff
@@ -60,7 +64,8 @@ public BodyGenerator(ModelGenerator modelGenerator, DataGenerator dataGenerator)
         }
     });
 
-    public String generate(String name, boolean isArray) {
+    public String generate(String name, boolean isArray, List<String> refs) {
+        LOG.debug("Generate " + name);
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/openapi/generators/BodyGenerator.java]: Worth adding an if.

File: src/org/zaproxy/zap/extension/openapi/generators/DataGenerator.java [hunk 5]
```diff
@@ -123,7 +126,21 @@ public String generateValue(Property items, boolean isPath) {
                     // You'd hope there was a cleaner way to do this, but I havnt found it yet :/
                     if (rp.get$ref().startsWith("#/definitions/")) {
                         String defn = rp.get$ref().substring(14);
-                        return this.generators.getBodyGenerator().generate(defn, false);
+                        LOG.debug("Dereferencing definition: " + defn);
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/openapi/generators/DataGenerator.java]: Worth adding an if.

File: src/org/zaproxy/zap/extension/openapi/resources/help/contents/openapi.html (hunk 1 above)
Comment (thc202) [inline src/org/zaproxy/zap/extension/openapi/resources/help/contents/openapi.html]: `it's`?

Comment (kingthorin) [inline src/org/zaproxy/zap/extension/openapi/resources/help/contents/openapi.html]: good catch....there should be an apostrophe

**Comment by psiinon** [comment]
Hopefully fixed all of those issues.
Btw, hope to have more enhancements soon, so might not be worth releasing this add-on yet.

**Comment by psiinon** [comment]
arg, the validation of hostOverride is too string - it fails if its not supplied :( Will fix now...

**Comment by psiinon** [comment]
That should be fixed now too

**Comment by thc202** [comment]
Thank you!

**Comment by psiinon** [comment]
@kingthorin ping? :)

**Comment by kingthorin** [comment]
LGTM

**Pull request merged**